### PR TITLE
Remote Backend: Support latest cost-estimate API

### PR DESCRIPTION
**Pull request by thrashr888**
Updates the remote backend to use the latest cost-estimate API, which works slightly differently. There are some unrelated changes to the mocks due to the go-tfe version bump.

To test/run this:
- setup your config using the remote backend
- add an AWS instance in the config
- `XC_OS="darwin" XC_ARCH="amd64" make bin` (for darwin machines) in this branch in terraform
- `<your terraform dir>/pkg/darwin_amd64/terraform plan` in the root of your config
- output should show plan results, then cost estimation results

--------

```HCL
terraform {
  backend "remote" {
    hostname = "app.terraform.io"
    organization = "pthrasher_v2"
    token = "<your TFC user token>"

    workspaces {
      name = "tfe-remote-backend-example"
    }
  }
}

provider "aws" {
  region  = "us-east-1"
}

resource "aws_instance" "basic-A" {
  ami           = "ami-62131702"
  instance_type = "t3.nano"
}
```

### CODE REVIEW

File: backend/remote/backend_common.go [hunk 1]
```diff
@@ -227,6 +228,59 @@ func (b *Remote) parseVariableValues(op *backend.Operation) (terraform.InputValu
...
+	if err != nil {
+		return generalError("Failed to retrieve cost estimation", err)
+	}
+
+	msgPrefix := "Cost estimation"
+	if b.CLI != nil {
+		b.CLI.Output(b.Colorize().Color(msgPrefix + ":\n"))
+	}
+
+	// FIXME: we need to loop for this instead
```
Comment (thrashr888) [inline backend/remote/backend_common.go]: FIXME here. The run gets stuck on this. We need to poll for one of the statuses below.

**Comment by thrashr888** [comment]
We might want to add the "pending" status to go-tfe:

https://github.com/hashicorp/go-tfe/blob/master/cost_estimate.go#L37-L42

`CostEstimatePending  CostEstimateStatus = "pending"`

File: backend/remote/backend_common.go (hunk 1 above)
Comment (thrashr888) [reply backend/remote/backend_common.go]:
```Error: Failed to retrieve cost estimation: The struct field was not of a known number type

The configured "remote" backend encountered an unexpected error. Sometimes
this is caused by network connection problems, in which case you could retry
the command. If the issue persists please open a support ticket to get help
resolving the problem.```

File: backend/remote/backend_plan_test.go [hunk 2]
```diff
@@ -709,7 +709,41 @@ func TestRemote_planWithWorkingDirectoryFromCurrentPath(t *testing.T) {
...
+	}
+	if run.PlanEmpty {
+		t.Fatalf("expected a non-empty plan")
+	}
+
+	output := b.CLI.(*cli.MockUi).OutputWriter.String()
+	if !strings.Contains(output, "Running plan in the remote backend") {
+		t.Fatalf("expected remote backend header in output: %s", output)
+	}
+	if !strings.Contains(output, "Resources: 1 of 1 estimated") {
```
Comment (thrashr888) [inline backend/remote/backend_plan_test.go]: This is failing

File: backend/remote/backend_common.go [hunk 3]
```diff
@@ -227,6 +229,88 @@ func (b *Remote) parseVariableValues(op *backend.Operation) (terraform.InputValu
...
+				b.CLI.Output(b.Colorize().Color(fmt.Sprintf("Resources: %d of %d estimated", ce.MatchedResourcesCount, ce.ResourcesCount)))
+				b.CLI.Output(b.Colorize().Color(fmt.Sprintf("           $%s/mo %s$%s", ce.ProposedMonthlyCost, sign, deltaRepr)))
+
+				if len(r.PolicyChecks) == 0 && r.HasChanges && op.Type == backend.OperationTypeApply {
+					b.CLI.Output("\n------------------------------------------------------------------------")
+				}
+			}
+
+			return nil
+		case tfe.CostEstimatePending:
```
Comment (mwhooker) [inline backend/remote/backend_common.go]: this should  maybe be 

```
case tfe.CostEstimatePending, tfe.CostEstimateQueued:
```

so we give status on either case, right?

Comment (thrashr888) [reply backend/remote/backend_common.go]: Ah, right. I thought that was the syntax for covering both. That's just some bad go-ing, then. Will fix!

File: backend/remote/backend_common.go [hunk 4]
```diff
@@ -227,6 +229,88 @@ func (b *Remote) parseVariableValues(op *backend.Operation) (terraform.InputValu
...
+			return stopCtx.Err()
+		case <-cancelCtx.Done():
+			return cancelCtx.Err()
+		case <-time.After(500 * time.Millisecond):
+		}
+
+		// Retrieve the cost estimation to get its current status.
+		ce, err := b.client.CostEstimates.Read(stopCtx, r.CostEstimate.ID)
+		if err != nil {
+			return generalError("Failed to retrieve cost estimation", err)
```
Comment (davidcelis) [inline backend/remote/backend_common.go]: `:s/estimation/estimate/`

File: backend/remote/backend_common.go [hunk 5]
```diff
@@ -227,6 +229,88 @@ func (b *Remote) parseVariableValues(op *backend.Operation) (terraform.InputValu
...
+			if b.CLI != nil && (i == 0 || current.Sub(updated).Seconds() > 30) {
+				updated = current
+				elapsed := ""
+
+				// Calculate and set the elapsed time.
+				if i > 0 {
+					elapsed = fmt.Sprintf(
+						" (%s elapsed)", current.Sub(started).Truncate(30*time.Second))
+				}
+				b.CLI.Output(b.Colorize().Color("Waiting for cost estimation to complete..." + elapsed + "\n"))
```
Comment (davidcelis) [inline backend/remote/backend_common.go]: `:s/estimation/estimate/`

File: backend/remote/backend_common.go [hunk 6]
```diff
@@ -227,6 +229,88 @@ func (b *Remote) parseVariableValues(op *backend.Operation) (terraform.InputValu
...
+			}
+			continue
+		case tfe.CostEstimateErrored:
+			return fmt.Errorf(msgPrefix + " errored.")
+		case tfe.CostEstimateCanceled:
+			return fmt.Errorf(msgPrefix + " canceled.")
+		default:
+			return fmt.Errorf("Unknown or unexpected cost estimation state: %s", ce.Status)
```
Comment (davidcelis) [inline backend/remote/backend_common.go]: `:s/estimation/estimate/`

File: backend/remote/backend_plan_test.go [hunk 7]
```diff
@@ -59,7 +59,7 @@ func TestRemote_planBasic(t *testing.T) {
 		t.Fatalf("expected remote backend header in output: %s", output)
 	}
 	if !strings.Contains(output, "1 to add, 0 to change, 0 to destroy") {
-		t.Fatalf("expected plan summery in output: %s", output)
+		t.Fatalf("expected plan summary in output: %s", output)
```
Comment (davidcelis) [inline backend/remote/backend_plan_test.go]: nice

File: backend/remote/backend_common.go (hunk 4 above)
Comment (thrashr888) [reply backend/remote/backend_common.go]: thanks. coulda done this as a suggestion and got commit credit! :P

File: backend/remote/backend_mock.go [hunk 8]
```diff
@@ -960,6 +1050,14 @@ func (m *mockWorkspaces) Read(ctx context.Context, organization, workspace strin
 	return w, nil
 }
 
+func (m *mockWorkspaces) ReadByID(ctx context.Context, workspaceID string) (*tfe.Workspace, error) {
```
Comment (mwhooker) [inline backend/remote/backend_mock.go]: this stuff looks unrelated

File: backend/remote/backend_common.go (hunk 3 above)
Comment (mwhooker) [reply backend/remote/backend_common.go]: for some reason I thought the same thing when we were working on this

File: backend/remote/backend_mock.go (hunk 8 above)
Comment (thrashr888) [reply backend/remote/backend_mock.go]: Yeah but it's required for valid mocks. I had to update these from upgrading the version of go-tfe.

File: backend/remote/backend_common.go [hunk 9]
```diff
@@ -227,6 +229,87 @@ func (b *Remote) parseVariableValues(op *backend.Operation) (terraform.InputValu
...
+			// Check if 30 seconds have passed since the last update.
+			current := time.Now()
+			if b.CLI != nil && (i == 0 || current.Sub(updated).Seconds() > 30) {
+				updated = current
+				elapsed := ""
+
+				// Calculate and set the elapsed time.
+				if i > 0 {
+					elapsed = fmt.Sprintf(
+						" (%s elapsed)", current.Sub(started).Truncate(30*time.Second))
```
Comment (mikenomitch) [inline backend/remote/backend_common.go]: Does this give the unit of time or will it just say "(30 elapsed)"?

Comment (thrashr888) [reply backend/remote/backend_common.go]: It's copied from the waitForRun code. `current.Sub(started).Truncate(30*time.Second)` returns whatever the rest of terraform does (`30s`, `12m`, etc)
